Editorial review: Document Wasm bulk memory operations (#44296)
mdn/content: files/en-us/webassembly/reference/definitions/elem/index.md

---
title: "elem: Wasm definition"
short-title: elem
slug: WebAssembly/Reference/Definitions/elem
page-type: webassembly-instruction
browser-compat: webassembly.definitions.elem
sidebar: webassemblysidebar
---

> The **`elem`** [definition](/en-US/docs/WebAssembly/Reference/Definitions) declares a series of function references that can be copied into a Wasm [`table`](/en-US/docs/WebAssembly/Reference/Definitions/table).

Reviewer: I think it might be worth making an analogy to data segments here. This is what I have usually found effective when explaining it to people. For example:

```suggestion
The **`elem`** [definition](/en-US/docs/WebAssembly/Reference/Definitions) declares an *element segment*, which is series of references that can be copied into a Wasm [`table`](/en-US/docs/WebAssembly/Reference/Definitions/table). They provide a way to initialize a table on instantiation, analogous to [data segments](/en-US/docs/WebAssembly/Reference/Definitions/data) for Wasm [memories](/en-US/docs/WebAssembly/Reference/Definitions/memory).
```

Also note that I struck "function"; tables and element segments can hold any kind of reference.

Author: That sounds good to me. Added in my next commit.

> 

Reviewer 2: This is good. Do you also want to mention the functionality provided by the declarative form here? - since that has nothing to do with tables you can't predict its purpose or that this form my exist from the current text.

Reviewer: In my opinion it's not worth mentioning because declarative segments are such a strange hack.

Author: I'm going to leave it.

{{InteractiveExample("Wat Demo: elem", "tabbed-taller")}}

```wat interactive-example
(module
  ;; table with 2 function slots
  (table $return_values 2 funcref)

  ;; Define functions
  (func $f1 (result i32)
    i32.const 42
  )
  (func $f2 (result i32)
    i32.const 100
  )

  ;; initialize table slots actively

>   (elem (i32.const 0) func $f1 $f2)

Reviewer: ```suggestion
  (elem (table $return_values) (offset (i32.const 0)) func $f1 $f2)
```

Again I'd suggest using the un-abbreviated form. The `func $f1 $f2` thing is also technically an abbreviation for the following:

```
(elem (table $return_values) (offset (i32.const 0)) (ref func) (ref.func $f1) (ref.func $f2))
```

In this case, `(ref func)` is the type of the table (non-nullable func references), and `(ref.func $f1)` etc. are the constant expressions returning the actual function references. The `func $f1 $f2` thing is terse enough, though, and common enough, to probably be fine for this example imo.

Tables can also be used to store other kinds of references, but I doubt it's worth having examples of this here. Functions are still by far the most common thing to put in a table.

Author: OK, as said above, I've updated it to the equivalent form as used in the `data` page.

We can always expand it at a later date if we feel like more syntax details are needed.

>   (elem (table 0) (offset i32.const 0) func $f1 $f2)

Reviewer 2: Given you went to the effort of naming the table `$return_values` why use `(table 0)`?

Author: Updated to

```
(elem (table $return_values) (offset i32.const 0) func $f1 $f2)
```

## Syntax

```plain
;; Active form, table initialized immediately
elem name table_identifier offset value_type element_list

;; Passive form, initialized later via table.init
elem name value_type element_list

;; Declaration form, declares already existing reference(s)
elem name declare value_type element_list
```

- `elem`
  - : The `elem` definition type. Must always be included first.
- `name` {{optional_inline}}
  - : An optional identifying name for the elem. This must begin with a `$` symbol, for example `$my_table`. If this is omitted, the `elem` can be identified (for example when calling `elem.drop`) by its index, for example `0` for the first `elem` in the wasm script, `1` for the second, etc.
- `table_identifier` {{optional_inline}}
  - : An identifier representing the `table` instance to place the table elements into. This can be one of:
    - `name`
      - : An identifying name [set for the `table`](/en-US/docs/WebAssembly/Reference/Definitions/table#name) when it was first defined. This must begin with a `$` symbol and be preceded by a `table` keyword, for example `(table $my_table)`.
    - `index`

>       - : An [`i32`](/en-US/docs/WebAssembly/Reference/Value_types/i32) value representing the index number of the table, for example `0` for the first table in the module, `1` for the second, etc.

Reviewer: Ditto the identifier notes above. Both `(table 1)` and `(table $foo)` are allowed, and `(table ...)` is required by the spec.

Author: Updated

- `offset` {{optional_inline}}
  - : An integer representing the offset at which to start placing the elements into the `table`. This is written as a Wasm [`i32`](/en-US/docs/WebAssembly/Reference/Value_types/i32) value (for example `(i32.const 0)`), and is only included if the `table_index` is written in `(table num)` form.
- `declare` {{optional_inline}}

>   - : A keyword that identifies the `elem` definition as being of the declarative form, meaning that it identifies function references so they can be referenced later on, with them being inserted in a `table`.

Reviewer: ```suggestion
  - : A keyword that identifies the `elem` definition as being of the declarative form, meaning that it identifies function references so they can be referenced later on, with them being inserted into a `table`.
```

Author: Reading this back, it didn't make much sense to me, so I updated it to:

> A keyword that identifies the `elem` definition as being of the declarative form, meaning that it declares function references that will be used at runtime (for example, by `ref.func`), without them being inserted into a table.

- `value_type` {{optional_inline}}
  - : A value type that defines which type of reference will be stored in this table. All references in the `element_list` must match this type. The value can be:
    - `func`: Function names, for example `$my_func`.
    - [`funcref`](/en-US/docs/WebAssembly/Reference/Value_types/funcref): Full function references, for example `(ref.func $my_func)`.
    - [`externref`](/en-US/docs/WebAssembly/Reference/Value_types/externref): External value references, for example `(ref.null extern)`.

>       Defaults to `func`.

Reviewer: This is all wrong. Tables can hold references; a table type can be any reference type ([syntax definition](https://webassembly.github.io/spec/core/syntax/types.html#syntax-reftype), [text format definition](https://webassembly.github.io/spec/core/text/types.html#reference-types)). This means tables can hold function references, externrefs, GC object references (structs and arrays), exception references, and so on. Function references are by far the most common, but this definition will accept all of the following and more: `funcref` / `(ref null func)` (nullable function references), `(ref func)` (non-nullable function references), `anyref` / `(ref null any)` (nullable references to GC values), `externref` / `(ref null extern)` (nullable references to host values, e.g. JS objects), and so on.

Notably absent from this is `func`. `func` is used for a special-purpose abbreviation that makes it more concise to declare a list of non-nullable function references. These two lines are precisely equivalent:

```
(elem (ref func) (ref.func 0) (ref.func 1) (ref.func 2))
(elem func 0 1 2)
```

The important thing to note is that `func` is _not_ a real WebAssembly reference/value type; it is just a keyword used for elem segments to indicate "what follows is a list of bare indices, please assume they are function indices". Without the `func` keyword, parsers will expect constant expressions, meaning you actually have to type `(ref.func ...)` for each function. As a result, this is _not_ valid:

```
(elem (ref func) 0 1 2)
```

Concrete things to fix:
- `funcref` and `externref` are not the only valid types; all reference types are allowed.
- `func` is part of a special case syntax that subsumes both the value type and element list.
- Nothing about this "defaults to func" in any meaningful way.

Author: OK, quite a lot to unpack here. I've updated this section to hopefully fix this, but you should read through carefully to see if what I've done is accurate. Quite a lot of very similar syntactic terms here.

- `element_list`
  - : A space-separated list of function references indicating the function references to be stored in the `table`. These can take the form of:
    - A specific function reference, for example `(ref.func $my_func)`, or the shorthand version that lists the function name only, for example `$my_func`. In the shorthand version, the function name is implicitly wrapped in `(ref.func)`.
    - A null `funcref`, for example `(ref.null func)`.

>     - A null `externref`, for example `(ref.null extern)`.

Reviewer: Per above, `$f1` is not really a shorthand for `(ref.func $f1)`. `(ref.func $f1)` is a constant expression producing a value of type `(ref func)` (non-nullable function reference). $f1 is just the index of a function, and for it to be used here, you must use the `func 0 1 2 ...` shorthand, which subsumes both the value type and element list.

And again, identifiers are literally nothing more than sugar for a bare index. Do not over-index on "names"; you can always replace an $identifier with an index.

You should also remove the bits about null `funcref`s and `externref`s, or at least not imply that the list is exhaustive. Other valid expressions you could use to get a ref include `(global.get $myref)`, `(struct.new $mystruct)`, `(ref.null exn)`, etc.

Author: OK, I've updated the wording before the list to

> The value can be any reference type, such as:

So as not to suggest that the list is exhaustive.

For the syntax stuff, should I just remove the examples from here? Are you saying that if you use function names (rather than the bare index numbers), you need to use the func shorthand?

Reviewer: All this element segment stuff is truly the worst, both in the text and binary encodings. There are waaaay too many options. I think I will try to write up a bit of a short guide this morning and get back to you.

## Description

> Wasm `elem` definitions define a series of function references that can be copied into a previously-defined [`table`](/en-US/docs/WebAssembly/Reference/Definitions/table). There are three forms of `elem` definition that you should know about:

Reviewer: Tables and element segments can hold any kind of reference, not just function references.

Author: At this point, I went through and changed all the language that makes it sound like tables store exclusively function references.

- [Active form](#active_form)
- [Passive form](#passive_form)
- [Declarative form](#declarative_form)

### Active form

> In active form, as seen in the live example earlier on, a table first needs to be defined:

Reviewer: I think it would be worth clarifying the purpose of the active form at the beginning:

```suggestion
An active element segment will be written into its table on instantiation. In active form, as seen in the live example earlier on, a table first needs to be defined:
```

Author: Added

>     - `nullref`, `nullfuncref`, `nullexternref`: Null references.

Reviewer 2: Consider making this a definition list.

Author: done

> Wasm `elem` definitions define a series of references that can be copied into a previously-defined [`table`](/en-US/docs/WebAssembly/Reference/Definitions/table). There are three forms of `elem` definition that you should know about:

Reviewer 2: ... unless there are some you are hiding from me!

```suggestion
Wasm `elem` definitions define a series of references that can be copied into a previously-defined [`table`](/en-US/docs/WebAssembly/Reference/Definitions/table). There are three forms of `elem` definition:
```

Reviewer 2: And again, declarative form has nothing to do with tables, which makes that first sentence a bit constraining.

I'd be tempted to leave sentence above alone and have a summary sentence on each of Active, Passive, Declarative below.

Author: The suggestion is good. I've done that, but then I also updated it further to solve the second issue you brought up. I ended up with:

> Wasm `elem` definitions define a series of references. There are three forms of `elem` definition:

I moved the "a previously-defined [`table`](/en-US/docs/WebAssembly/Reference/Definitions/table)" bit to the active form description.

```wat
(table $return_values 2 funcref)
```

You then define the functions you want to store references to in the `table`, and declare an `elem` definition that includes the function references.

```wat
(func $f1 (result i32)
  i32.const 42
)
(func $f2 (result i32)
  i32.const 100
)

(elem (i32.const 0) func $f1 $f2)
```

> This `elem` definition includes the `value_type` to be stored (`func`), and the `element_list` to store in the table (`$f1 $f2`). Most significantly, it includes a number indicating the index of the table (`table_index`) to store the functions in — `(i32.const 0)` — which indicates the first table in the module.

Reviewer: `(i32.const 0)` is the index _in the table_ where data should be written. It does NOT specify _which_ table to write into. For example:

```wat
(module
  (table 4 funcref)
  (table 4 funcref)
  (func (result i32) i32.const 42)
  (func (result i32) i32.const 100)
  (elem (i32.const 2) func 0 1)
  (elem (table 1) (i32.const 1) func 0 1)
)
```

This module declares two tables, and initializes them with two functions, but writes those two function references at different offsets. Table 0 will have values `null, null, <func 0>, <func 1>`, while table 1 will have values `null, <func 0>, <func 1>, null`.

Note that we selected the second table in the second element segment by using `(table 1)`, NOT by doing `(i32.const 1)`. The first element segment has an implicit `(table 0)` before the `(i32.const 2)`.

Author: Crikey, you can see how confused I was getting at this point ;-)

OK, rewritten so that it makes more sense.

Optionally, you can include an `offset` value to specify the table element number you want to store the first function in. In such a case, the `table_index` is written in a different form (`(table 0)`) and the `offset` is written as an `i32` value:

```wat
(elem (table 0) (i32.const 1) func $f1 $f2)

> ```

Reviewer: Hopefully you can see why this is slightly wrong. This example does indeed write to table 0 starting at index 1, but it is not true that "the table index is written in a different form". In the previous paragraph, the `(i32.const 0)` referred to index 0, and the `(table 0)` was implicit.

Maybe one way to clear this up is: in the active form of both data segments and element segments, both memory/table index and starting offset are required, because the runtime needs to know exactly how to write the data on instantiation. However, if referring to memory/table 0, you can omit the explicit `(memory 0)` or `(table 0)` in the text format. You can NEVER omit the starting offset, however. Therefore you should not say it is "optional" here.

Author: I just got rid of this bit. With the above bit corrected, it was no longer necessary.

In this case, functions `$f1` and `$f2` would be stored starting at table element `1`.

### Passive form

> In passive form, the `elem` definition declares the functions that should be stored in the table in the same way as in active form. The main difference is that, in passive form, you don't specify the `table_index` or optional `offset` value. This means that the function references are not stored in the `table` immediately. Instead, this part of the process is handled manually using a [`table.init`](/en-US/docs/WebAssembly/Reference/Table/init) instruction.

Reviewer: ```suggestion
In passive form, the `elem` definition declares the functions that should be stored in the table in the same way as in active form. The main difference is that, in passive form, you don't specify the `table_index` or `offset` value. This means that the function references are not stored in the `table` immediately. Instead, this part of the process is handled manually using a [`table.init`](/en-US/docs/WebAssembly/Reference/Table/init) instruction.
```

Author: updated

Let's see what this looks like in code. We include the `elem` definition in a similar manner to the active form example, except that this time we don't include the `table_index`. Instead, we include a `name` value (`$funcs`) to identify the `elem` later on.

```wat
(elem $funcs funcref (ref.func $f1) (ref.func $f2))
```

We can then call `table.init`, referencing the `elem` `name`, to copy those functions into the specified table:

```wat
(func (export "init")
  i32.const 0    ;; destination table index
  i32.const 0    ;; offset into elem segment
  i32.const 2    ;; number of elements to copy
  table.init $funcs
)
```

You can see a full working example at [Passive `elem` example](#passive_elem_example).

### Declarative form

> The declarative form of `elem` is needed when you want to use `ref.func` in your code without putting the function into a table. It allows you to create a reference to that function that can be referenced via `ref.func`:

Reviewer: This is fine, but on its own I'm not sure it clarifies why you would need such a thing. (And I don't blame you for any confusion here, because this was a huge point of confusion for me when I was learning Wasm.) Take the following as a deeper dive into the weeds rather than something we need to teach on MDN.

Basically I think the non-obvious part is "it allows you to create a reference to that function that can be referenced via `ref.func`". You would naturally think that you could reference any function with `ref.func` at any time. But this is not true; inside of WebAssembly functions, you can actually only do `ref.func` on any function that has in some way been marked "refable". The way you do this is by referencing the function in some other part of the module besides the body of a function, such as in a `global`, an `import`, or an `elem` segment. So in this example, some functions can be ref'd at runtime, while others cannot:

```wat
(module
  (import "foo" "bar" (func $imported))
  (func $f1 (param i32))
  (func $f2 (param i64))
  (func $f3 (param f32))
  
  (global funcref (ref.func $f1))
  (elem declare funcref (ref.func $f2))

  (func (export "getImported") (result funcref)
    ref.func $imported ;; valid, imports are refable
  )
  (func (export "getF1") (result funcref)
    ref.func $f1 ;; valid, $f1 was used in a global declaration
  )
  (func (export "getF2") (result funcref)
    ref.func $f2 ;; valid, $f2 was used in an elem segment
  )
  (func (export "getF3") (result funcref)
    ref.func $f3 ;; invalid, $f3 was never used outside a function body
  )
)
```

Why is this? To my knowledge, it's because if a function can be called indirectly (e.g. via `call_indirect` or `call_ref`), then engines have to generate more code for that function, for implementation reasons. It is therefore useful to know up front which functions may be called indirectly, and which may not; any functions that cannot be called indirectly can produce less machine code. However, if `ref.func` was always allowed, then you would have to assume function 0 could be ref.func'd by function 99999 somewhere down the line. So _before_ function compilation starts, you want to know which funcs are refable and which are not.

Therefore the designers of wasm decided that any function that appears in a module outside the function body would be refable, and all others would not be. In the binary format, the bodies of functions come (almost) dead last, so this just works.

And so finally, the purpose of `elem declare` is to allow you to just dump a bunch of function indices that you want to be refable down the line, with no further side effects. Nothing will be exported, nothing will be saved to a global, nothing will be added to a table. It's just this "useless" little element segment that exists only to be a big list of refable functions.

Author: Thanks for the explanation. This makes sense now. I have added a short paragraph that explains why this is needed, after the code snippet.

```wat
(module

>   Create a reference to the $add function

Reviewer: ```suggestion
  ;; Create a reference to the $add function
```

Author: Done, thanks

  (func $add (param i32 i32) (result i32)
    local.get 0
    local.get 1
    i32.add
  )

  (func (export "getRef") (result funcref)
    ;; only valid because of the declarative elem above
    ref.func $add
  )
)
```

This was added to the language because normally you can only reference functions with `ref.func` that have been made referencable, for example in a [`global`](/en-US/docs/WebAssembly/Reference/Definitions/global) definition or by being imported from the JavaScript host. Declarative `elem` definitions exist to make some functions referencable that otherwise wouldn't be.

## Examples

### Passive `elem` example

> This example shows how you can use the passive form of `elem` to not immediately copy the specified references to the table, instead copying them later on via the [`table.init`](/en-US/docs/WebAssembly/Reference/Table/init) instruction.

Reviewer 2: immediately copy feels a bit clunky. Maybe something like

```suggestion
This example shows how you can use the passive form of `elem` to defer copying the specified references to the table on instantiation, later adding them using the [`table.init`](/en-US/docs/WebAssembly/Reference/Table/init) instruction.
```

Reviewer 2: NOTE, through all this it is very clear what the options are, but no so clear why I might want to defer vs instantiate immediately. Only a "FWIW" because it is probably obvious to people using this.

Author: Your wording suggestion is much better — used it.

As for why you might want to do this, I've not got a clear answer for that, but I think it is one of those fundamental things, akin to why you sometimes immediately declare a JS variable and initialize it with a value, and sometimes don't store the value in it until later on.
